### ISSUE TITLE
Introducing esm.run - A New-Age CDN for JavaScript modules

### ISSUE THREAD

**Issue by MartinKolarik** [jsdelivr/jsdelivr#18263]
**We are excited to announce a long-awaited feature - improved support for packages distributed as ES modules 🎉**

It is currently in beta, which means we encourage you to try it out, but not use it in production applications yet. This issue will cover all technical details and can be also used to provide feedback or report bugs if you find any.

# TL;DR

1. Demo and essential information: https://www.jsdelivr.com/esm
2. You can load any ESM file like this:
   - `https://cdn.jsdelivr.net/npm/uuid/+esm` (default file)
   - `https://cdn.jsdelivr.net/npm/uuid/dist/esm-browser/index.js/+esm` (explicit file).
3. There's a new domain that you can use for shorter and prettier links. It redirects everything to our main domain and is not meant to be used in production:
   - `https://esm.run/uuid`
   - `https://esm.run/uuid/dist/esm-browser/index.js`

# How this works

## Locating entry points

To find the default file, we use `module` and `jsnext:main` fields from `package.json`. These fields are currently used by all existing tooling and by most package authors. It is very likely we'll also add support for the [node.js `exports` field](https://nodejs.org/api/packages.html#packages_main_entry_point_export) in the future.

### Resolving

1. Absolute URLs like `https://example.com/foo` are always kept in their original form.
2. Relative imports are first resolved via the [`browser` field](https://github.com/defunctzombie/package-browser-field-spec) if it exists (this applies to the entry point as well) and then resolved using [node's experimental resolution algorithm](https://nodejs.org/api/esm.html#esm_resolution_algorithm), which supports extension resolution and importing from directories that include an index file. The supported extensions are: `.mjs`, `.js`, and `.json`.

### Bundling

1. Internal imports (pointing to files in the same package):
   - are bundled with the requested file if they are static, e.g., `import foo from './foo.js'`,
   - are rewritten to their jsDelivr URL if they are dynamic, e.g., `import('./foo.js')`.

2. External imports (pointing to files in other packages):
   - are always rewritten to their jsDelivr URL, whether they are static or dynamic.

This means that one npm package roughly equals one bundle and one HTTP request, which allows good caching and doesn't require too many HTTP requests.

External URLs are always generated with fully resolved dependency versions based on `package.json` dependencies (or latest if the version is not specified there). This further improves caching and guarantees stable versions even for transitive dependencies.

## Performance

 - The generated files are automatically minified.
 - Source maps are available for easy debugging.
 - Once generated, the files are stored in our permanent storage and served from there.
 - In supported browsers, HTTP preloading is used to start fetching external static dependencies as soon as possible.

# What doesn't work (but may in the future)
 - Packages that are written in CJS, or a mix of ESM and CJS.
 - Packages that import core node.js modules.
 - Packages that import files in formats other than JS/JSON.
 - Packages larger than 50 MiB.
 - search on esm.run lists all packages. This will soon be changed to list only the supported ones.

# The esm.run domain

This domain is introduced as an easier to remember and type alternative. It doesn't run on our multi-cdn architecture and simply redirects all requests to the main cdn domain, so it's not meant to be used in production.

**Comment by Finesse** [comment]
Is there a URL option to bundle all imports inside a single file? Comments in files produced by jsDeliver say that Rollup is used so it won't be a problem.

**Comment by evromalarkey** [comment]
Will the current URL scheme change to me more friendly? Currently what I see in devtools/network is bunch of +esm files

I guess this `https://cdn.jsdelivr.net/npm/d3@6.3.1/+esm/d3.js` would be better than this `https://cdn.jsdelivr.net/npm/d3@6.3.1/+esm`

Skypack and esm.sh have similar approach for production urls
https://cdn.skypack.dev/pin/react@v17.0.1-tOtrZxBRexARODgO0jli/min/react.js
https://cdn.esm.sh/react@17.0.1/esnext/react.js

**Comment by MartinKolarik** [comment]
@evromalarkey I did notice this problem but you can easily configure devtools to show full paths by right-clicking the column:

![image](https://user-images.githubusercontent.com/6192491/102700520-891a0c00-424e-11eb-999c-5e378b720cab.png)

Alternatively, you could also add the filename as a query string, which we ignore: https://cdn.jsdelivr.net/npm/d3@6.3.1/+esm?d3

**Comment by FredKSchott** [comment]
Congrats on the release! I've run your benchmarks myself a few times now, and can't reproduce your numbers across a few different devices. Here's what I see on my laptop:

<img width="843" alt="Screen Shot 2021-01-25 at 7 41 36 PM" src="https://user-images.githubusercontent.com/622227/105797838-a628b580-5f45-11eb-9157-7ed6c5421af2.png">

I don't know if this is out of date or just bad, but the data appears outdated across both Skypack and https://esm.run.

Additionally, your benchmark imports are to https://cdn.jsdelivr.net/npm/d3/+esm, and not https://esm.run. That's fine if you want to measure the "optimized" use-case, but then that should really be compared against Skypack's optimized URL as well: https://docs.skypack.dev/skypack-cdn/api-reference/pinned-urls-optimized

Can you update these stats to better represent things to your users?

**Comment by MartinKolarik** [comment]
Hi @FredKSchott, the performance may of course change over time as each of the services develops. We would have preferred showing realtime results but that wasn't possible so we went with numbers that were accurate at the time this was launched. We'll recheck and update if the results are consistently different now. 

Regarding the pinned URLs, those are not functionally equivalent because they lock down a specific version while jsDelivr (even on our primary domain) and unpkg resolve to latest.

**Comment by vp2177** [comment]
I'm not sure this is the appropriate place to report this, I've tried this service with `lit-element` and while there are no errors logged, the `html` function doesn't seem to work

**Comment by MartinKolarik** [comment]
Thanks for the report @vp2177. I don't immediately see where's the problem as `L.html` returns the correct object but we'll check this further later.

**Comment by vp2177** [comment]
Well the issue is that it renders `[object Object]` but after simply loading `lit-element` from SkyPack instead of JsDelivr, no other changes, it renders as it should (renders a `<h2>Lit</h2>`). 


I think `html` is re-exported by `lit-element` from `lit-html`, and this looks like an issue with how that import from `lit-html` is rewritten

**Comment by MartinKolarik** [comment]
@vp2177 after closer inspection we found the problem but not yet sure how we'll be able to fix this.

lit-element imports main lit-html file [here](https://github.com/Polymer/lit-element/blob/master/src/lit-element.ts#L63) and it also imports another nested file `shady-render.js` from lit-html [here](https://github.com/Polymer/lit-element/blob/master/src/lit-element.ts#L57). The problem is `shady-render.js` imports the main lit-html file as well.

When we get a request for `shady-render.js` we include all of its imports, including the main lib-html file (this is by design - as explained in the first post here, internal imports are always bundled).

Then when lib-element imports the main lib-html file in a separate request, it's loaded a second time. This would not be a breaking problem in most cases but the library relies on `instanceof` checks and those don't work if two versions of the code are loaded.

**Comment by MartinKolarik** [comment]
@FredKSchott it seems Skypack is now consistently faster than it was before so we decided to update the numbers.

**Comment by evromalarkey** [comment]
https://esm.run/@fullcalendar/daygrid fails to load, you should check all @fullcalendar plugins though, I think it's because import css (eg. https://cdn.jsdelivr.net/npm/@fullcalendar/timegrid/main.js)

Another problem is https://esm.run/dayjs (works on https://esm.run/dayjs/esm though)

**Comment by evromalarkey** [comment]
Also I see a potential problem with resolving subfolders (only with production url) when using importmaps

```json
{
  "imports": {
    "dayjs": "https://cdn.jsdelivr.net/npm/dayjs/esm/+esm",
    "dayjs/": "https://cdn.jsdelivr.net/npm/dayjs/esm/+esm/"
  }
}
```
`import isBetween from "dayjs/plugin/isBetween.js"` - this wouldn't work

it would work with esm.run url though
```json
{
  "imports": {
    "dayjs": "https://esm.run/dayjs/esm",
    "dayjs/": "https://esm.run/dayjs/esm/"
  }
}
```

It would work with `https://cdn.jsdelivr.net/npm/dayjs/esm/`, but that's only because this library has esm version builded on npm. The same problem could be in many other libraries with subfolders.

**Comment by MartinKolarik** [comment]
> I think it's because import css

Yes, that is not supported yet, as mentioned in the original post.

> Another problem is https://esm.run/dayjs (works on https://esm.run/dayjs/esm though)

dayjs removed the "module" package.json entry in the recent versions - I'm not sure why but there isn't much we can do in this case.

Regarding the import maps - will check this closer.

**Comment by mikabytes** [comment]
Would it be possible to add HTTP2 push support, and an option to disable bundling?

With bundling, we get this issue:

Imagine having `A.js` imports `B.js` and `C.js` and a SPA app. On initial fetch only `C.js` is imported as the rest aren't needed. Later on, A.js is imported.

Result:

- `C.js` has been downloaded twice
- `C.js` has been parsed and evaluated twice
- The two imports of C.js are _not_ the same. They are completely different files, so references are broken

Bundling is dinosaur performance thing. We better avoid it when possible.

**Comment by MartinKolarik** [comment]
@mikabytes indeed, that is basically the issue described in https://github.com/jsdelivr/jsdelivr/issues/18263#issuecomment-774369214 and we might end up introducing some options to fix it.

**Comment by Xeevis** [comment]
Hello, I have trouble importing libraries that use `process.env.NODE_ENV` to differentiate development and production builds. For example

```js
import tippy from "https://cdn.jsdelivr.net/npm/tippy.js/+esm"; 
```
throws
```diff
- ReferenceError: process is not defined
- at https://cdn.jsdelivr.net/npm/tippy.js/+esm:12:2691
```

In snowpack non-production code (including the expression) is stripped away.

> https://cdn.skypack.dev/-/tippy.js@v6.3.1-HeICNxUmzNoshMxx5q78/dist=es2020,mode=imports/optimized/tippyjs.js
(no `process.env.NODE_ENV`)

vs

>https://cdn.jsdelivr.net/npm/tippy.js@6.3.1/+esm 
(plenty of `process.env.NODE_ENV`)

Since esm.run is using Rollup, this should be easy to solve with [@rollup/plugin-replace](https://www.npmjs.com/package/@rollup/plugin-replace) plugin (which is also what Tippy [suggests](https://atomiks.github.io/tippyjs/v5/faq/#i-m-getting-uncaught-referenceerror-process-is-not-defined)).

```js
import replace from '@rollup/plugin-replace';

export default {
  plugins: [
    replace({
      'process.env.NODE_ENV': JSON.stringify('production')     
    })
  ]
};
```

**Comment by MartinKolarik** [comment]
We're aware of this pattern for CJS files but I'd say it's a bad idea to assume `process` exists in ESM code. One of the key advantages of ESM is that it should work in a browser without bundling so the check is usually written in a safe way as `process && process.env && process.env.NODE_ENV`

Still, if some modules rely on it, we'll look into replacing `NODE_ENV` to improve the compatibility.

**Comment by zarianec** [comment]
@MaximKing1 this is regular benchmark results from https://www.jsdelivr.com/esm. You just need to resize your browser's width until the page gets redrawn for smaller devices.

**Comment by IgorNovozhilov** [comment]
> @vp2177 after closer inspection we found the problem but not yet sure how we'll be able to fix this.
> 
> lit-element imports main lit-html file [here](https://github.com/Polymer/lit-element/blob/master/src/lit-element.ts#L63) and it also imports another nested file `shady-render.js` from lit-html [here](https://github.com/Polymer/lit-element/blob/master/src/lit-element.ts#L57). The problem is `shady-render.js` imports the main lit-html file as well.
> 
> When we get a request for `shady-render.js` we include all of its imports, including the main lib-html file (this is by design - as explained in the first post here, internal imports are always bundled).
> 
> Then when lib-element imports the main lib-html file in a separate request, it's loaded a second time. This would not be a breaking problem in most cases but the library relies on `instanceof` checks and those don't work if two versions of the code are loaded.

Found a similar problem in [`Material Web Components`](https://github.com/material-components/material-web).
[`mwc-tab-bar-base.ts`](https://github.com/material-components/material-web/blob/6278ee5df98ca7eae0d694e58b052fdbf4b0519d/packages/tab-bar/mwc-tab-bar-base.ts#L18) imports file [`mwc-tab-base.ts`](https://github.com/material-components/material-web/blob/6278ee5df98ca7eae0d694e58b052fdbf4b0519d/packages/tab/mwc-tab-base.ts) as `import {TabBase} from '@material/mwc-tab/mwc-tab-base';`
At the same time file `mwc-tab-base.ts` imported as nested file in [`mwc-tab.ts`](https://github.com/material-components/material-web/blob/6278ee5df98ca7eae0d694e58b052fdbf4b0519d/packages/tab/mwc-tab.ts#L12), which is also imported from `mwc-tab-bar-base.ts` as `import {Tab} from '@material/mwc-tab';`.
And as a result, this check does not work: [`.filter((e: Node) => e instanceof TabBase`](https://github.com/material-components/material-web/blob/6278ee5df98ca7eae0d694e58b052fdbf4b0519d/packages/tab-bar/mwc-tab-bar-base.ts#L78), because the class declaration is duplicated

## Proposal

There is a suggestion how to solve this problem via `package.json#exports`

1. _(For unambiguity of import and restoring order)_ For "+esm" mode, if `package.json` contains `exports` block external import from CDN on the path to the file not specified in `exports`

> Like in CommonJS, module files within packages can be accessed by appending a path to the package name unless the package’s package.json contains an "exports" field, in which case files within packages can only be accessed via the paths defined in "exports".
> (с) https://nodejs.org/dist/latest-v16.x/docs/api/esm.html#esm_terminology

2. Export paths from `exports` always provide the independent URI's, and replace the import of nested files to inside a bundled file to external paths, if they are finally resolved in the ways specified in `exports`

For example, for the above case
Both files (https://cdn.jsdelivr.net/npm/@material/mwc-tab/+esm, https://cdn.jsdelivr.net/npm/@material/mwc-tab-bar/+esm) will contain:
`import{TabBase as c}from"/npm/@material/mwc-tab@0.25.1/mwc-tab-base/+esm";`

Perhaps this will be another headache for modules authors , but it will develop people's desire for a thoughtful declaration of dependencies and exported paths for the module. `imho` :cat2: 

**FAQ**
Anticipating the question that this will increase the number of downloaded files, I will answer that this problem is solved module authors by generalizing the export to one point. It is their decision whether to allocate independent access points, or leave a single file.
> I can notice that even directly in "Node.js", importing from different export points can periodically cause problems with duplicating code downloads with different versions. (Tested by personal bitter experience :exploding_head:)

**Comment by MartinKolarik** [comment]
@IgorNovozhilov the problem is definitely fixable by package authors writing the imports in a different way but our goal is to make as many packages as possible write without requiring specific patterns.

**Comment by IgorNovozhilov** [comment]
> make as many packages as possible work

In this case, if you just add replacement nested file imports to external ones, it will definitely increase the number of packages that will work correctly when importing from your CDN :upside_down_face:


> without requiring specific patterns

`package.json#exports` is gradually becoming more popular, it has much more possibilities for use in combination with ES modules. In my opinion, I would still consider `exports` as basic definition of available export points, and not as specific pattern for "+esm" mode

**Comment by claviska** [comment]
That's fair and would 100% satisfy my concerns. Thanks for clarifying!

**Comment by MartinKolarik** [comment]
@Xeevis we shipped the `process.env.NODE_ENV` replacement today. It doesn't apply to previously-generated files so far, but works with the new ones, e.g. https://cdn.jsdelivr.net/npm/tippy.js@6.3.3/+esm

**Comment by ylemkimon** [comment]
If there are no dependencies in the ESM, there should be no difference from the original jsDelivr URLs except minification, right? For instance, are https://cdn.jsdelivr.net/npm/katex@0.15.1/dist/katex.mjs and https://cdn.jsdelivr.net/npm/katex@0.15.1/dist/katex.mjs/+esm identical except minification?

**Comment by zarianec** [comment]
@ylemkimon optimizations (like tree-shaking) may apply to the bundled code so the content could vary. But even with optimizations functionality should be identical to original files.

In your particular case, the bundled file is identical to the original one.

**Comment by cristiano-belloni** [comment]
Is the resolving / bundling code open source?

**Comment by MartinKolarik** [comment]
@cristiano-belloni no but it's mostly based on rollup.

**Comment by MartinKolarik** [comment]
We made many improvements here in the past weeks. For now, all of them are available at our preview domain: https://jsdelivr-origin-dev.herokuapp.com/npm (some are already on the production version too):

1. Importing CSS: should work now, see e.g., https://cdn.jsdelivr.net/npm/@fullcalendar/daygrid/+esm
2. `process.env.NODE_ENV` is now replaced with `production` and dead code elimination is applied.
3. `package.json` exports/imports fields are supported.
4. CJS modules are supported.

We're still considering the following issues:

1. Friendlier URL scheme.
2. Import maps prefix resolving.

**Comment by nihgwu** [comment]
I wanted to try `esm.run` over `esm.sh` in after [this PR](https://github.com/nihgwu/react-runner/pull/102), it's super fast and most of the examples work fine, expect [MUI](https://play-react.vercel.app/#JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQokscAggG7IzJRxEnkABEAFdgAehCcsUYMgA2Y9p260G4aPABCwmDAgA7XsVJkhoiVJnyx23QdWMNcAMKpUmtFjnB9WI-1MRcUkYaVkFV3dPb18HdWYAFSwADxgAMWAvABN-EzNgy3CxJNSM7LimeDToEGcDGGI5ABlkACMvXMEgi1CrBWqoWvrGlva5CqdnAAssXABrVohkzsDzELDradmFpYnmJp85lfyejYUD-Tm9+ABxGRy+PO71vrE74CzruE0l4+fC6w-ZJfJoQeYAeR0MSwWQAkrgDH9zMAEfpUABaF5FUEQqE+GFfBIAT0gAHMoMgwFMiUiCr0isSyRSqUSvnV9Fx8TxHl01gCIvVkFyvgBvOC4ShSBIzEBYAA0cGlWFlAAViGwPtI4ABfWmnV6oGBEuRYVCqOgEYT6fDARF1MBEmSkqYwAAUYGIYFQAC4UPoiQBKOAiuhwOCUGDCKCGV2hsNwAA8jIg5Mp1LgHCsHIAvAAiRZZIkAJlz4ogcmgedCqQAdKhZgYstwiaX5MBSfo87gsBzpKWRTXBx6IF7tQA+OPx4Nke2O9suuAASvI2sn8YTFyOCIrUDzPhmMhgpamlAIeZdMC93rEElENYRIDEuYnU6nAE0IFG4AB1LCtVDAKEa5hgmYibmOIpkCuwHBr4ADucAACJSK6AY1qSWDpMIchyG+OBQKhq6vtONZkER67FCSKbMtSL5wAG9CrnQqKGnAMAyn42bipKoRKrKqH0HQKSVHAWRYAQyDYYQVo2oiADK7b6LC+iocGcYsfAUzIPoWQmnJwitCAgFwFxrpYGwPYwL62B4DANYDCAACiFkcgmAASCQALJNA5jkmrKHJjkG2ZjmpU7mZZNYehFHKIeJklyG6DFrhponKCZcDwXADnIVwZkuXZuBRpQHIJNwGEwMlU43nApoxDA6JZMAqBtCa6K+Kk6LQplEDoix5ZYClBioANNYVqSrohsRypCnIvpNlw6GYa6ZAzcA4wBnKMFgGgqBwdAWTzcoS1umQO1uPtUCfJta7alVOqCWGEZRjGa5JhxaoQBqYk8OxypYNmIp-bK44wQm7Kcr4PAPpAvg5rmkg+KWkjJN+HzsXmySoM+MEgZEHj1t1Yh0cRCZArj8aoMkgNTcRr6SFApI+AkI6+gAHFtdOvk1qBgHIyBEr6ZAECawKc1z8YiykiHAJQsn6EL27CCA+hkOLEsoN4HawqEIA+uQ3a9lAasU2G2rkcRJN0wmShcDwVM03AIC+gAjAqrSktu0BC-WqJNlARI1ojqs6qDGsgTicyQjA0JwqicDE6biaKBwdtW6TyapiyZbqHDMDni7paZrI8NTAArDj4dwApHZwD4SegZnNFEunr5k78MMGJZeYEDUpYGPphmAYDWk6XpBlGTAur6BAABqbYLX4DsimKIBWXALuh63GcpOkmRyFkSdTgzTOdrmM+DPIuZH-GlAAI6iJQh9V2Glo4WjWTsTfYYfHma1yNfF+cB+ZjDzI5RGchWBZCyJQNwgCX76GQLKP+ED4FV0knoOo4ATShBQbNNBEtE5VyTLvMoB9v5O3Kj4PMF9JAAIoffR+MIKFvzkB-L+QDEHINzOdPaB0CEaxAV4PMKpdqXSyAIiWRowAAx4WI-hFDf5yIugooBGCIBYL5phWRRUoAlQarw8RkjiJEPDgmByEMRhtC8BQ1EDRyyA3BjMeYixlgcDkMIWRlBZQgHaFAUsXtdw8JkAzFsCcxwW3DkIuQeZsA+L8U7LAxjXymI1mTHQeh9AUOkbI1AE9ALJOIqw9hUwKHF20vnXMdihS+AkRQ5eq914AGYFS+N9EWUON9t50xroYZSN9QK2Eyd0qcCZ3g5GqVyEZbdxl111nALG0zSabjgCecSeYADERduAl0qQWYslcgHxgGKSCA8BDEHQAPwUJAmBQ4SyKLjIeSBWZgFlTPPXCstZZ5cxbIzDsipeZ9klg+VOEUuZEL2E0sgCyfoUC4ARFaGAlzq6KTgAAVTALmSJxC7mXA+aBJ5Ay3j3GmYMpY3TyXJEpbOJ0C4GlO3XhzJ2rRfQABZQ7hLBmICGNTpBW0bh9dUmooB0WStqIAA), you can reproduce the issue [here](https://jsfiddle.net/c73xrvLd/), I'd say SkypackCDN has the same issue as well, but somehow esm.sh fixed it

**Comment by MartinKolarik** [comment]
Thanks for reporting @nihgwu! Turns out we didn't properly purge the old files versions in some cases. This should be fixed soon.

**Comment by nihgwu** [comment]
@MartinKolarik forgot to mention another issue, I got two requests for the same react version, which will cause [this error](https://reactjs.org/docs/error-decoder.html/?invariant=321), I double checked the request url is exactly the same, don't know why, unless I explicitly use `https://cdn.jsdelivr.net/npm/react@17.0.2/+esm` and `https://cdn.jsdelivr.net/npm/react-dom@17.0.2/+esm`

**Comment by nihgwu** [comment]
@MartinKolarik thank you for your quick response, really appreciated. It almost works, now I got `TypeError: Cannot read properties of null (reading 'createSvgIcon')` error, it will work if I don't import the icon, [reproduce demo](https://jsfiddle.net/n1zgs3v6/)

And [this issue](https://github.com/jsdelivr/jsdelivr/issues/18263#issuecomment-1067192639) is still there, and there is another issue `SyntaxError: The requested module 'https://esm.run/react/jsx-runtime' does not provide an export named 'Fragment'`

**Comment by MartinKolarik** [comment]
> And https://github.com/jsdelivr/jsdelivr/issues/18263#issuecomment-1067192639 is still there, which is the biggest problem for me, I can't use https://cdn.jsdelivr.net/npm/react@17.0.2/+esm as alias as then jsx-runtime will be resolved to https://cdn.jsdelivr.net/npm/react@17.0.2/+esm/jsx-runtime which is wrong, and I don't want to change all of my react import to remote version as in that way I won't get typings

I don't follow here, I don't see any duplicate requests in the jsFiddle.

> It almost works, now I got TypeError: Cannot read properties of null (reading 'createSvgIcon') error, it will work if I don't import the icon, [reproduce demo](https://jsfiddle.net/n1zgs3v6/)

Indeed, I don't see why yet, if you got a chance to pinpoint what exactly is going wrong and where it would greatly help (there's many nested imports so we need to find which file exactly is getting transformed incorrectly if it's supposed to work).

**Comment by nihgwu** [comment]
@MartinKolarik I tried to workaround the dup react issue [here](https://github.com/nihgwu/react-runner/pull/104/files#diff-ae1897249ca83a6393a91971111e48e87c641e055d125c3441fd3079e363e311R21), and I think probably you should use the same strategy, and in that way we will get better debug experience, right now they are all `+esm` in the network panel, so I propose the following format
```
// response for `https://esm.run/react`
export * from 'https://cdn.jsdelivr.net/npm/react@17.0.2/+esm'
export { default } from 'https://cdn.jsdelivr.net/npm/react@17.0.2/+esm'
```

**Comment by nihgwu** [comment]
@MartinKolarik Regarding the duplicated requests for React, [repro demo](https://jsfiddle.net/fm5ja19s/) in jsFiddle

The first and forth requests are importing the same content, and for my case, the request urls are also the same
![image](https://user-images.githubusercontent.com/2595058/158280778-8ffea196-9d2d-4fdb-aebe-c90149d85d0a.png)

**Comment by nihgwu** [comment]
@MartinKolarik And for the icon issue, [here](https://jsfiddle.net/oujqpzkw/) is the problem

**Comment by MartinKolarik** [comment]
> @MartinKolarik Regarding the duplicated requests for React, [repro demo](https://jsfiddle.net/fm5ja19s/) in jsFiddle
> 
> The first and forth requests are importing the same content, and for my case, the request urls are also the same ![image](https://user-images.githubusercontent.com/2595058/158280778-8ffea196-9d2d-4fdb-aebe-c90149d85d0a.png)

I believe here it would help if you loaded `https://esm.run/react@17.0.2` in your code but the versions handling is indeed a little problematic, I'll see if we can make it more intuitive.

**Comment by MartinKolarik** [comment]
> @MartinKolarik And for the icon issue, [here](https://jsfiddle.net/oujqpzkw/) is the problem, the import result is `null`

The file doesn't have a default export even in the original version so that seems right: https://cdn.jsdelivr.net/npm/@mui/material@5.5.1/utils/index.js

**Comment by MartinKolarik** [comment]
Indeed, we may need to do the reexport thing for this case.

**Comment by MartinKolarik** [comment]
Also regarding the paths, there's a reasonable workaround https://github.com/jsdelivr/jsdelivr/issues/18263#issuecomment-748531613 but we're still considering other options too.

**Comment by nihgwu** [comment]
> Indeed, we may need to do the reexport thing for this case.

Yes I just tested, it's not only for react but any package with a dependent, the dependency will always be requested twice, we should reexport to solve it

**Comment by JosefJezek** [comment]
I found issue with Lit lib.

https://cdn.jsdelivr.net/npm/lit@2.2.3/+esm

https://jsfiddle.net/startweb/ko73cjaq/

File `lit-element@3.2.0/lit-element.js` imports older `reactive-element@1.3.0` and `lit-html@2.2.0`

https://cdn.jsdelivr.net/npm/lit-element@3.2.0/lit-element.js/+esm

**Comment by jimmywarting** [comment]
When i try to import `cbor-x` that have some optional `node:buffer` support to speed up certain things then it likes to inline nodejs buffer into the package. 

Here is a cjs example that you can see do not have any `require('buffer')` syntax
https://cdn.jsdelivr.net/npm/cbor-x@1.4.1/dist/index.js

but when i add `/+esm` then it includes the Buffer module which it dose not need to function correctly. Buffer is optional
https://cdn.jsdelivr.net/npm/cbor-x@1.4.1/dist/index.js/+esm

I really dislike that NodeJS added the Buffer module onto the global namespace it the first place... it should have been just like any other core module... ppl should really be using `import Buffer from 'node:buffer'` if they intend to use it...

is there any way to circumvent this? i don't want or need the buffer module...

**Comment by MartinKolarik** [comment]
This is intentional because, unfortunately, many packages rely on the global Buffer and don't work without it. If a global Buffer reference is found it the code, we add the polyfill.

**Comment by MartinKolarik** [comment]
The package could opt-out from this by using an `import Buffer from 'node:buffer'` and then excluding this import from browser builds via `exports` or `browser` fields.

**Comment by jimmywarting** [comment]
can i (they) do something like `const Buffer = globalThis.Buffer` at the top of the file? or even possible to something like

-`const Buffer = globalThis['Buffer']` or
- `const Buffer = globalThis['Buf'+'fer]`?

**Comment by MartinKolarik** [comment]
Importing it as a module and then adding this to `package.json` is the best solution ([docs](https://github.com/defunctzombie/package-browser-field-spec)):
```
"browser": {
    "node:buffer": false
}
```

Alternatively https://nodejs.org/dist/latest-v19.x/docs/api/packages.html#subpath-imports can be used and for browser it can resolve to an empty file.

`const Buffer = globalThis['Buf'+'fer]` would likely work too, just not as clean as the other solutions.

**Comment by jimmywarting** [comment]
just hope that it don't do minifications first and resolve `'Buf'+'fer'` into `'Buffer'` before it tries to polyfill stuff. but i will pass your suggestion along to the other repo, thanks for your help!

**Comment by mon-jai** [comment]
Will the `esm.run` server be open sourced?

**Comment by porfirioribeiro** [comment]
Support for dependency version?

I am currently aliasing vue and vue-i18n to esm.run
```
vue: "https://esm.run/vue@3",
"vue-i18n": "https://esm.run/vue-i18n@9",
```

But this will fail, as `vue@3` resolves to `vue@3.2.47`, but `vue-i18n` imports `vue@3.2.45`
That means it will download 2 versions of vue and will not work.

Of course i could change my dependency to be `vue@3.2.45`, and that will work, but it will break as soon as i have other dependency that requires `vue@3.2.46` for example.

`esm.sh` solves this by adding a `dep` option to allow specifying a specific version.

```
vue: "https://esm.sh/vue@3",
"vue-i18n": "https://esm.sh/vue-i18n@9?deps=vue@3",
```

Having something like this in `esm.run` would be really nice

**Comment by Potherca** [comment]
Playing around with `/+esm`, I noticed that it does not work together with `/combine`. 

The error I get is:

> Uncaught SyntaxError: Identifier 't' has already been declared (at +esm:30:20)

<sup>(Where `t` is sometimes `e` or another letter)</sup>

Is this a bug, or is combining ESM not supported?

**Comment by MartinKolarik** [comment]
Yes, combine for esm is not supported as it would quite easily result in conflicts in the exports.

**Comment by Potherca** [comment]
When I try to `/combine/` different formats, I get an error message that  this isn't allowed.

For clarity, it might be worth adding a similar message that `/+esm` is disallowed?

**Comment by midnight-wonderer** [comment]
I came to report an issue only to find out that a lot of people had already reported it.
Basically, whenever two versions of the same library are loaded, the integration is broken.

For me
~~~javascript
import { h, render } from 'https://cdn.jsdelivr.net/npm/preact@10.24.3/+esm';
import { signal } from 'https://cdn.jsdelivr.net/npm/@preact/signals@1.3.0/+esm';
~~~
broke. (Both are the latest versions.)

This is because `https://cdn.jsdelivr.net/npm/@preact/signals@1.3.0/+esm` integrates itself with an older version of preact esm, loaded internally. Render the integration useless.

Any ideas for moving forward? Or simply don't load shared resources in such cases.

**Comment by jlukic** [comment]
I'm seeing issues when using css imports from jsdelivr using css imports. They appear to link to the correct files but then are loaded like JS modules causing the following error
![Image](https://github.com/user-attachments/assets/15ca44b5-f081-434b-87db-b5b6f4e391fe)

You can see it live here https://next.semantic-ui.com/examples/tailwind 

This example works locally when the packages are streamed through Vite in dev mode (which supports the [`?raw` qs param](https://vite.dev/guide/assets#importing-asset-as-string) but fails when loaded from CDN

Is there any guide to best practices to getting css loaded?

----

Also its worth noting on a separate note, it might make sense to disambiguate in the docs whether the entrypoint for jsdelivr is the 'jsdelivr' export condition or a top level field labeled 'jsdelivr'. This [part of your docs is still a bit confusing for me](https://www.jsdelivr.com/documentation#id-configuring-a-default-file-in-packagejson)

**Comment by tats-u** [comment]
Is there a plan to migrate from the good old Rollup + terser to Rolldown? It should be faster and reduce the computing resource.
